### PR TITLE
[syncd] Enable bulk api for neighbor entries

### PR DESCRIPTION
**Pull request by Ndancejic**
SAI 1.11.0 added support for bulk neighbor entries. Adding support for neighbor bulk operations to syncd.

needed for ado: #25072867

### CODE REVIEW

**Comment by kcudnik** [comment]
looks good, make code coverage and unnitests, and move out from draft then i will review it

**Comment by Ndancejic** [comment]
@kcudnik Could you take a look? not sure what else I need for coverage, but the tests that are failing are common among other PRs out right now

**Comment by kcudnik** [comment]
not passing tests you can ignore, those are flaky, but you can see that you don't have tests that cover your code, and here is exactly showed where: https://dev.azure.com/mssonic/build/_build/results?buildId=552818&view=codecoverage-tab

**Comment by kcudnik** [comment]
```
main::test_bulk_neighbor: Fresh start
Killing syncd
Flushing redis
Starting syncd
Replay BCM56850/bulk_neighbor.rec
terminate called after throwing an instance of 'std::runtime_error'
  what():  :- translate_local_to_redis: failed to translate local RID oid:0x20000000006ee
player BCM56850/bulk_neighbor.rec: exitcode: 134:
FAIL: BCM56850.pl
```

you can make run tests locally on your dev machine by making "make check" and post when they pass

RID translation is problem that you used RID value of object that was not created previously in script or was not obtained by GET api

**Comment by Ndancejic** [comment]
> ```
> main::test_bulk_neighbor: Fresh start
> Killing syncd
> Flushing redis
> Starting syncd
> Replay BCM56850/bulk_neighbor.rec
> terminate called after throwing an instance of 'std::runtime_error'
>   what():  :- translate_local_to_redis: failed to translate local RID oid:0x20000000006ee
> player BCM56850/bulk_neighbor.rec: exitcode: 134:
> FAIL: BCM56850.pl
> ```
> 
> you can make run tests locally on your dev machine by making "make check" and post when they pass
> 
> RID translation is problem that you used RID value of object that was not created previously in script or was not obtained by GET api

I run into the following error when trying to run autogen.sh: `configure.ac:32: error: AM_COND_IF: no such condition "CODE_COVERAGE_ENABLED"`

**Comment by kcudnik** [comment]
you need to install " autoconf-archive" packaged in ubuntu that contains that code coverage sctipt

**Comment by Ndancejic** [comment]
> you need to install " autoconf-archive" packaged in ubuntu that contains that code coverage sctipt

I see, I'll get that set up for next time then. I was able to get the coverage passing in the meantime, so this PR is ready for review/merge

File: configure.ac [hunk 1]
```diff
@@ -12,7 +12,6 @@ AC_PROG_LIBTOOL
 AC_HEADER_STDC
 AM_PATH_PYTHON3
 AX_PYTHON3_DEVEL
-AX_CODE_COVERAGE
```
Comment (prsunny) [inline configure.ac:15]: Why is this changed in this PR?

**Review by prsunny — COMMENTED** [review]
Kamil, could you please review/signoff?

File: configure.ac (hunk 1 above)
Comment (Ndancejic) [reply configure.ac:15]: Good catch, I was trying to make the build work locally and it snuck into my PR, I'll revert this change and update

Comment (kcudnik) [reply configure.ac:15]: please revert

File: configure.ac [hunk 2]
```diff
@@ -29,10 +28,6 @@ AM_CONDITIONAL(DEBUG, test x$debug = xtrue)
 
 AM_COND_IF([DEBUG], AC_SUBST(DBGFLAGS, "-ggdb -DDEBUG"), AC_SUBST(DBGFLAGS, "-g"))
 
-AM_COND_IF([DEBUG],
```
Comment (kcudnik) [inline configure.ac:32]: please revert

File: syncd/Syncd.cpp [hunk 3]
```diff
@@ -1574,6 +1616,27 @@ sai_status_t Syncd::processBulkSetEntry(
...
+                entries[it].rif_id = m_translator->translateVidToRid(entries[it].rif_id);
+            }
+
+            status = m_vendorSai->bulkSet(
+                    object_count,
+                    entries.data(),
+                    attr_lists.data(),
+                    mode,
+                    statuses.data());
+
```
Comment (kcudnik) [inline syncd/Syncd.cpp:1636]: extra empty line not needed

File: syncd/tests.cpp [hunk 4]
```diff
@@ -671,6 +671,147 @@ void test_bulk_route_set()
...
+    sai_object_id_t switch_id;
+    status = sai_switch_api->create_switch(&switch_id, 1, &swattr);
+
+    ASSERT_SUCCESS("Failed to create switch");
+
+    std::vector<std::vector<sai_attribute_t>> neighbor_attrs;
+    std::vector<const sai_attribute_t *> neighbor_attrs_array;
+    std::vector<uint32_t> neighbor_attrs_count;
+
+    for (uint32_t i = index; i < index + count; ++i)
```
Comment (kcudnik) [inline syncd/tests.cpp:717]: this is wired ? from 15 to 15+count ? simplify to 0 to count

File: syncd/tests.cpp [hunk 5]
```diff
@@ -671,6 +671,147 @@ void test_bulk_route_set()
...
+    std::vector<sai_status_t> statuses(count);
+    status = sai_neighbor_api->create_neighbor_entries(count, neighbors.data(), neighbor_attrs_count.data(), neighbor_attrs_array.data(), SAI_BULK_OP_ERROR_MODE_IGNORE_ERROR, statuses.data());
+    ASSERT_SUCCESS("Failed to create neighbor");
+    for (size_t j = 0; j < statuses.size(); j++)
+    {
+        status = statuses[j];
+        ASSERT_SUCCESS("Failed to create neighbor # %zu", j);
+    }
+
+    for (uint32_t i = index; i < index + count; ++i)
```
Comment (kcudnik) [inline syncd/tests.cpp:774]: this is confusing

**Comment by Ndancejic** [comment]
@kcudnik Coverage and tests are passing in the last commit, could you please review

**Comment by kcudnik** [comment]
@Ndancejic  i made big refactor on master, removed a lot of unnecessary code, please resolve conflicts and check build

**Comment by Ndancejic** [comment]
@kcudnik rebased onto the latest branch, please take a look and approve if no additional comments

**Review by kcudnik — APPROVED** [review]

**Comment by kcudnik** [comment]
```
./unittest/vslib/test_sai_vs_neighbor.cpp
./unittest/lib/test_sai_redis_neighbor.cpp
```
those files are no longer needed, interface was refactored, and they are not connected into Makefile, can be removed

**Comment by dprital** [comment]
> @Ndancejic can you please create a separate cherry-pick to `202405` branch? We see Dual-ToR test failures due to missing implementation

@Ndancejic , Do we have PR for 202405 ?

**Comment by bingwang-ms** [comment]
Cherry-pick is done by #1415.
